90% - Prevent null values being added to ExtendedGraph

lordtatty opened the pull request:
To address issue https://github.com/talis/tripod-php/issues/81

This will ensure that subjects, predicates, and resource values are always strings.  Literal values can be any scalar - we would like to enforce strings for literal values too, but doing so may break legacy databases which use non-strings as values.

lordtatty commented on src/classes/ExtendedGraph.class.php:
```diff
@@ -171,24 +171,40 @@ public function add_literal_triple($s, $p, $o, $lang = null, $dt = null) {
+     *
+     * @return bool
+     */
+    protected function isValueValid($value){
```
Wouldn't we want to check resource values as well though?  ``isValidTripleValue``?

rsinger replied:
From Tripod's perspective, don't we just need to ensure that subject, predicate, and object are all strings?

rsinger replied:
non-empty strings

lordtatty replied:
Agree, but enforcing strings as opposed to null could be very painful for any graphs which are already storing non-string values (eg. integers).  This current PR to ignore null values could well cause enough headaches without taking out integers at the same time.   I do think we should ultimately enforce strings-only, but maybe one step at a time, once we are comfortable with no nulls.

rsinger replied:
Tripod *only* deals in strings, does it not?

lordtatty replied:
It should only deal with strings, but up until now there has been nothing to enforce it...  Clearly this method will be the right place to enforce that - but is it what we want to do right now?

rsinger replied:
Ok, so strings and integers.

rsinger replied:
Or arrays of strings and integers.

rsinger replied:
I guess what I'm saying is that we should be doing type checking here so we're not just allowing some other bad data that is equally as incompatible as a null value.

lordtatty replied:
How do we deal with legacy data, though?  We have no idea what crazy things have ended up in the graph.  The php data types are:

```
String.
Integer.
Float (floating point numbers - also called double)
Boolean.
Array.
Object.
NULL.
Resource.
```

Integer, floats and booleans could all have ended up in the graph.  Is it possible to add an array via the public ExtendedGraph methods?

Agree that really we should be ignoring anything that isn't a string, but I guess the question is how much we are prepared to break anything currently using tripod to enforce this.

rsinger replied:
Tripod only supports plain literals.  Full-stop.  If other data types are in the documents they would produce invalid RDF, anyway, because we do not support typed literals at all.

rsinger commented on src/classes/ExtendedGraph.class.php:
```diff
@@ -171,24 +171,40 @@ public function add_literal_triple($s, $p, $o, $lang = null, $dt = null) {
+     * @return bool
+     */
+    protected function isValidTripleValue($value){
+        if(!is_string($value) && !is_int($value) && !is_array($value) && !is_float($value) && !is_bool($value)){
```
Why not ` is_scalar ` here?

lordtatty replied:
Uh yep, I meant to use that, no idea why I didn't.... one moment.

lordtatty commented:
@rsinger @robotrobot Something I just noticed is https://github.com/talis/tripod-php/blob/master/src/classes/ExtendedGraph.class.php#L141 - ``@return boolean true if the triple was new, false if it already existed in the graph``.

What should be return of ``add_resource_triple`` and ``add_literal_triple `` be if we aren't adding the triple due to the value being invalid?  Null?  False?  At the moment it is false.

lordtatty commented:
@rsinger @robotrobot also at the moment we are allowing scalars for both literals and resources. Should we restrict resources to only strings? I really can't imagine how having anything other than a string can be legitimate for a uri.

rsinger commented on src/mongo/MongoGraph.class.php:
```diff
@@ -138,39 +142,65 @@ private function add_tarray_to_index($tarray)
+     *
      * @param array $mongoValueObject
-     * @return array
+     * @return array| false an array of values or false if the value is not valid
```
`array| false` isn't a valid phpdoc, it needs to be `array|bool`

rsinger commented:
This looks good.  Should we validate `$s` and `$p` while we're here?  Is it possible to add invalid values for them currently?

lordtatty commented:
@rsinger  I feel like maybe that should be raised as a separate issue so we can get this branch down.

rsinger commented:
Why can't you just run $s and $p through isValidResourceProperty?

rsinger commented:
I mean, my real question is, can you even add invalid values there?  I don't see why wouldn't fix it here, if so.

lordtatty commented:
@rsinger my only real reservation to fixing that in this branch is that I'll have to add a whole bunch of tests to cover that, and this needs to get shipped today, ideally.  It's currently passing as it is so I think the right strategy is to merge this down and fix up the $s and $p in another branch...

rsinger commented:
Can you just confirm whether or not it's even an issue?  There is no point in making a release to prevent us from adding duff data if we've still left it open to add duff data.

lordtatty commented:
Added checks for subjects and predicates

lordtatty commented:
This will now also throw an exception if trying to add a triple with a blank subject

rsinger commented on src/classes/ExtendedGraph.class.php:
```diff
@@ -171,24 +177,60 @@ public function add_literal_triple($s, $p, $o, $lang = null, $dt = null) {
+     *
+     * @return bool
+     */
+    protected function isValidTripleValue($value){
```
Triple value?

lordtatty replied:
I thought that would be a good generic name to cover subject, predicate and value.

rsinger replied:
I think it made more sense when it was named resourcevalue

lordtatty replied:
It's used for more than just validating a resource value though - the only thing it doesn't validate now is a literal value.  ``isValidTripleComponentValue``?

lordtatty replied:
Should I be breaking this out into ``isValidSubject``, ``isValidPredicate``, and ``isValidVaue``?  It might make sense since we are now ensuring that the subject is not an empty string, as we can just add the check into there...

rsinger replied:
No, they're all just uris.

lordtatty replied:
I see your point, in my head I was separating subjects and uri values unnecessarily here, will fix up.

lordtatty replied:
I hadn't spotted that duplicate validation - that was a leftover from a previous iteration.  Removing.

lordtatty commented on src/classes/ExtendedGraph.class.php:
```diff
@@ -171,24 +177,60 @@ public function add_literal_triple($s, $p, $o, $lang = null, $dt = null) {
-            if ( ! in_array( $o_info, $this->_index[$s][$p] ) ) {
-                $this->_index[$s][$p][] = $o_info;
+        // Make sure the subject is not an empty string
+        if($s === ""){
```
I was just mooting that with Ross in a comment further down. Wondering if I should break up ``isValidTripleValue`` into more specific validation methods for the subject, predicate and value.

lordtatty replied:
Will do that shortly.

lordtatty commented on src/mongo/MongoGraph.class.php:
```diff
@@ -128,8 +128,24 @@ private function add_tarray_to_index($tarray)
-                $predicate = $this->qname_to_uri($key);
-                $predObjects[$predicate] = $this->toGraphValueObject($value);
+                // Make sure the predicate is valid
+                if($this->isValidTripleValue($key)){
```
That's true, but it will throw an error rather than an exception if the predicate is an empty string.  I've added ``testAddTripodArrayContainingEmptyPredicate`` to assert that a php error is thrown in this circumstance.  It's not a perfect test, but I think it's probably worth it to not have to unnecessarily validate the predicate.

lordtatty commented:
@robotrobot @rsinger Fixed up from PR comments:  
  - Renamed isValidTripleValue back to isValidResourceValue, and added the empty string check into there.  
  - Removed an unnecessary predicate validation which is being covered by the the labeller, and added appropriate tests to cover that.  
  - toGraphValueObject no longer returns false if the array is empty.
